confidential-data-hub/storage/src/volume_type/rbd/mod.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2024 Intel
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+pub mod error;
+
+use super::SecureMount;
+use anyhow::Context;
+use async_trait::async_trait;
+use base64::Engine;
+use error::{RbdError, Result};
+use log::{debug, error};
+use rand::{distributions::Alphanumeric, Rng};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use strum::{Display, EnumString};
+use tokio::{
+    fs,
+    io::{AsyncReadExt, AsyncWriteExt},
+    process::Command,
+};
+
+/// LUKS encrypt storage binary
+const LUKS_ENCRYPT_STORAGE_BIN: &str = "/usr/local/bin/luks-encrypt-storage";
+
+#[derive(EnumString, Serialize, Deserialize, Display, Debug, PartialEq, Eq)]
+pub enum RbdEncryptType {
+    #[strum(serialize = "luks")]
+    LUKS,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Debug)]
+struct RbdParameters {
+    /// The device number, formatted as "MAJ:MIN".
+    #[serde(rename = "deviceId")]
+    pub device_id: String,
+
+    /// The encryption type. Currently, only LUKS is supported.
+    #[serde(rename = "encryptType")]
+    pub encrypt_type: RbdEncryptType,
+
+    /// Encryption key. If not set, generate a random 4096-byte key
+    #[serde(rename = "encryptKey")]
+    pub encrypt_key: Option<String>,
+
+    /// Indicates whether to enable dm-integrity.
+    #[serde(rename = "dataIntegrity")]
+    pub data_integrity: String,
+}
+
+pub(crate) struct Rbd;
+
+async fn random_encrypt_key() -> anyhow::Result<String> {
+    let mut buffer = vec![0u8; 4096];
+    rand::thread_rng().fill(&mut buffer[..]);
+    Ok(base64::engine::general_purpose::STANDARD.encode(&buffer))
+}
+
+async fn get_plaintext_secret(secret: &str) -> anyhow::Result<String> {
```

Comment:
I think I was expecting the encryption key to be a resource uri. Maybe having it be a sealed secret is better because that contains info about the KBS/HSM, integrity protection, etc., but so far we haven't used a sealed secret this way in any other parts of the CDH. Are we planning to adopt sealed secret format as a general wrapper around a resource uri?

cc: @Xynnn007

Reply:
I think they are for different scenarios. Given that we now support both KMS as a KBS backend (use KBS URI to index a key) and KMS as a CDH plugin (use Sealed Secret to index a key), we might need to support both. We can add a condition judgement here, if with `kbs://` prefix, use 
```rust
new_getter("kbs", ProviderSettings::default()).await?.get_secret(&secret, &Annotations::default())
```
if with `sealed.` prefix, use
```rust
secret::unseal_secret(secret.as_bytes()).await?
```
else, raise an error saying that illegal keyid format.

Thus we could support both inside CDH code. The consequence is both deploy model could be used and the user does need to worry about underlying logic.

Reply:
Ok, I think supporting both makes sense. Each scheme has some pros/cons. I wonder if we should do something similar for other features.